base_layer/blockchain/src/blockchainstate.rs

```diff
@@ -22,45 +22,169 @@
+        let block_chain_state = BlockchainState {
+            headers,
+            outputs,
+            kernels,
+            range_proofs,
+            store,
+        };
+
+        Ok(block_chain_state)
+    }
+
+    fn build_db() -> Result<LMDBStore, DatastoreError> {
+        let builder = LMDBBuilder::new();
+        builder
+            .set_mapsize(5)
+            .set_path("./storage/state/")
+            //create for headers mmr
+            .add_database(&"headers_mmr_checkpoints".to_string())
+            .add_database(&"headers_mmr_objects".to_string())
+            .add_database(&"headers_init".to_string())
+            //create for outputs mmr
+            .add_database(&"outputs_mmr_checkpoints".to_string())
+            .add_database(&"outputs_mmr_objects".to_string())
+            .add_database(&"outputs_init".to_string())
+            //create for range_proofs mmr
+            .add_database(&"range_proofs_mmr_checkpoints".to_string())
+            .add_database(&"range_proofs_mmr_objects".to_string())
+            .add_database(&"range_proofs_init".to_string())
+            //create for kernels mmr
+            .add_database(&"kernels_mmr_checkpoints".to_string())
+            .add_database(&"kernels_mmr_objects".to_string())
+            .add_database(&"kernels_init".to_string())
+            .build()
+    }
+
+    /// Will update the pruning horizon to the new value for the outputs data store
+    pub fn change_pruning_horizon(&mut self, new_pruning_horizon: usize) {
+        self.outputs.set_pruning_horizon(new_pruning_horizon);
     }
 
     /// This function  will process a new block.
-    /// Note the block must have been validated by the chainstate before.
-    pub fn process_new_block(&self, new_block: &Block) -> Result<(), StateError> {
-        self.validate_new_block(new_block)
+    /// Note the block is consumed by the function.
+    pub fn process_new_block(&mut self, mut new_block: Block) -> Result<(), StateError> {
+        self.validate_new_block(&new_block)?;
+        self.prune_all_inputs(&new_block)?;
+        // All seems valid, lets add the objects to the state
+        let mut drainer = new_block.body.outputs.drain(..);
+        while let Some(output) = drainer.next() {
+            self.range_proofs.push(output.proof.clone())?;
+            self.outputs.push(output.into())?;
+        }
+        self.kernels.append(new_block.body.kernels)?;
+        self.headers.push(new_block.header)?;
+        // lets check states
+        self.check_mmr_states()?;
+        self.check_point_state()
     }
 
     /// This function will validate the block in terms of the current state.
-    pub fn validate_new_block(&self, _new_block: &Block) -> Result<(), StateError> {
+    pub fn validate_new_block(&self, new_block: &Block) -> Result<(), StateError> {
+        // we assume that we have atleast in block in the headers mmr even if this is the genesis one
```

Comment:
This comment?

Reply:
thats to explain that the unwrap on the following line should never fail to unwrap:
headers.get_last_added_object().unwrap()

Reply:
he he - I meant that you should rewrite it because it looks like 2 comments merged into one and doesn't make sense